libpromises/pipes_unix.c

```diff
@@ -142,6 +169,113 @@ static pid_t CreatePipeAndFork(const char *type, int *pd)
 
 /*****************************************************************************/
 
+static pid_t CreatePipeAndFork(const char *type, int *pd)
+{
+    IOPipe pipes[2] = {{0}};
+    pipes[0].type = type;
+    
+    pid_t pid = GenericCreatePipeAndFork(pipes);
+    
+    pd[0] = pipes[0].pipe_desc[0];
+    pd[1] = pipes[0].pipe_desc[1];
+
+    return pid;
+}
+
+/*****************************************************************************/
+
+static pid_t CreatePipesAndFork(const char *type, int *pd, int *pdb)
+{
+    IOPipe pipes[2] = {{0}};
+    /* Both pipes MUST have the same type. */
+    pipes[0].type = type;
+    pipes[1].type = type;
+
+    pid_t pid =  GenericCreatePipeAndFork(pipes);
+
+    pd[0] = pipes[0].pipe_desc[0];
+    pd[1] =  pipes[0].pipe_desc[1];
+    pdb[0] = pipes[1].pipe_desc[0];
+    pdb[1] = pipes[1].pipe_desc[1];
+    return pid;
+}
+
+/*****************************************************************************/
+
+IOData cf_popen_full_duplex(const char *command, bool capture_stderr)
+{
+/* For simplifying reading and writing directions */
+#define READ  0
+#define WRITE 1
+    int child_pipe[2];  /* From child to parent */
+    int parent_pipe[2]; /* From parent to child */
+    pid_t pid;
+
+    fflush(NULL); /* Empty file buffers */
+    pid = CreatePipesAndFork("rt", child_pipe, parent_pipe);
```

Comment:
This part is a bit confusing. ``type`` is supposed to signal what kind of pipe you want, following ``fopen()`` semantics. Here you are creating a bidirectional pipe, but mode is still ``rt``. The correct way to specify it according to ``fopen()`` would be ``r+t``. Whether we want to keep this semantic is debatable, but regardless I think we should choose just one, so that we don't have this inconsistent situation where ``rt`` (signaling unidirectional) is specified, but in fact bidirectional is created.